components/script/dom/htmlinputelement.rs

```diff
@@ -1135,14 +1263,46 @@ impl HTMLInputElement {
         let shadow_tree = self.shadow_tree.borrow();
         Ref::filter_map(shadow_tree, |shadow_tree| {
             let shadow_tree = shadow_tree.as_ref()?;
-            let ShadowTree::Color(color_tree) = shadow_tree;
-            Some(color_tree)
+            match shadow_tree {
+                ShadowTree::Color(color_tree) => Some(color_tree),
+                _ => None,
+            }
         })
         .ok()
         .expect("UA shadow tree was not created")
     }
 
     fn update_shadow_tree_if_needed(&self, can_gc: CanGc) {
+        if self.input_type() == InputType::Text {
+            let text_shadow_tree = self.text_shadow_tree(can_gc);
+            let value = self.Value();
+
+            let placeholder_text = match value.is_empty() {
```

Comment:
I think we don't need to dynamically set the placeholder's text content.

We can use `:host(:placeholder-shown) #input-placeholder { visibility: hidden !important }` instead.

Reply:
Sure, Thanks!

We still need to change its content whenever placeholder is modified by JavaScript though. But I guess it is much better cache this.